### PR TITLE
WIP: cluster-wide service ports

### PR DESCRIPTION
**Pull request by justinsb**
Uploading now for discussion wrt whether we merge this into #5225 

For AWS ELB load balancers, it looks like we will need to assign a port to externally load balanced services (more discussion in #5225).  This patch adds cluster-wide port assignments to the API server; there is more work to be done to have the kube-proxy respect this assignment and avoid the range when choosing random ports.

Comments welcome.  In particular, I'm not sure on the naming of the PortalServicePort parameter, or on whether we want a separate PoolAllocator (I think yes, because we will be moving this pool logic to etcd when we go multi-master).

Maybe I should split PoolAllocator into its own PR (and refactor the IP allocator to use it also)

### CODE REVIEW

File: pkg/util/port_range.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package util
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+type PortRange struct {
+	MinInclusive int
```
Comment (thockin) [inline pkg/util/port_range.go]: Prefer to implement this as base + count - it's clear what is inclusive/exclusive and it's how I plan to implement port ranges for services :)

File: pkg/registry/service/rest.go [hunk 2]
```diff
@@ -402,3 +477,16 @@ func externalLoadBalancerNeedsUpdate(oldService, newService *api.Service) bool {
 	}
 	return false
 }
+
+// Determine if the service needs a cluster-wide port assigned
+func (rs *REST) needsServicePort(service *api.Service) bool {
+	if service.Spec.CreateExternalLoadBalancer {
+		balancer, ok := rs.cloud.TCPLoadBalancer()
```
Comment (thockin) [inline pkg/registry/service/rest.go]: why create a balancer just to throw it away?  Why not cloud.LoadBalancerNeedsHostPort() or something?

File: pkg/master/master.go [hunk 3]
```diff
@@ -200,6 +202,11 @@ func setDefaults(c *Config) {
 		}
 		c.PortalNet = portalNet
 	}
+	if c.PortalServicePorts.Size() == 0 {
+		defaultPortRange := util.PortRange{MinInclusive: 10000, MaxExclusive: 11000}
```
Comment (thockin) [inline pkg/master/master.go]: How could we NOT do this for clouds that don't require it?  Or is that just not worth the effort?  If NeedsClusterServicePort() is false, we shouldn't need anything here...

File: pkg/master/master.go [hunk 4]
```diff
@@ -72,14 +72,15 @@ import (
...
-	PortalNet         *net.IPNet
-	EnableLogsSupport bool
-	EnableUISupport   bool
+	Cloud              cloudprovider.Interface
+	EtcdHelper         tools.EtcdHelper
+	EventTTL           time.Duration
+	MinionRegexp       string
+	KubeletClient      client.KubeletClient
+	PortalNet          *net.IPNet
+	PortalServicePorts util.PortRange
```
Comment (thockin) [inline pkg/master/master.go]: Needs a better name - this isn't really about portals it's about proxies.  Suggest ProxyPorts or NodeProxyPorts or ProxyPortRange or something.

File: pkg/api/types.go [hunk 5]
```diff
@@ -898,7 +898,10 @@ const (
 )
 
 // ServiceStatus represents the current status of a service
-type ServiceStatus struct{}
+type ServiceStatus struct {
+	// If the service is assigned a cluster-wide port, it is set here
+	ServicePort int `json:"servicePort,omitempty"`
```
Comment (thockin) [inline pkg/api/types.go]: This has to be handled like Portal IP - it must be part of Spec.  Specifically, Status must be "observable" information not desired state.

Suggest going back to the name we had when this sort of existed originally - ProxyPort.

**Comment by thockin** [comment]
This is more concept review than code review, but a few things I had to call out.

Now a series of disconnected thoughts after reading this in no real order.

I see no "good" way out of this just "not terrible" ways.  No matter what range we pick, someone will want to use it for a pod HostPort, so we just need to fail gracefully.

Generically I think we need a "managed ports" range in which we think it is plausible to broker ports.  We probably need to cross-check this with plain old host ports.  If a pod asks for a host port in this range we should not give that same port to a service.  If we're going to do that, why not make it the whole port space?  I guess we don't want to hand out service ports in the low numbers, so it makes sense that we differentiate the allocation from the pool.  Something like `mp := PortRange{Base: 1, Count: 32768}; spa := servicePortAllocater{Pool: &mp, Window: servicePortRange}`.  NB: avoid the ephemeral range, which can be found in /proc/sys/net/ipv4/ip_local_port_range.  How do we ensure that is consistent on nodes?

Is there something more radical we can do that is less horrible?  What if we aimed ELB at VMs with HAProxy and used an enlightened HAProxy.  I'm worried that this is an endless rabbit hole.  Let's dicuss on VC.

**Comment by justinsb** [comment]
Mini break-through (maybe?): It looks like we can create our ELBs in a different VPC subnet.  And thus, we will know the IP range of the ELBs.  So, for example, I created a second subnet of 172.20.1.0/24, and the source IP of traffic through ELB was 172.20.1.44.  We still don't know which ELB is which, and thus we will still need a cluster-wide port assigned per externally-load-balanced-service (because ELB cannot target a different port per service), but I think we can avoid port collisions with kube-proxy's dynamic ports.

Here's a picture from the AWS docs:
http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/UserScenariosForVPC.html

My personal preference would still be to reserve a port range for these cluster-wide ports, and block kube-proxy and/or pod HostPorts from using them.  I think that opens up two use-cases: cluster federation via SRV records and cheaper load-balancing by marking a service as externally available (without provisioning a load balancer).  The only loss of functionality is for pod host-ports (which now can't use the reserved range), but I think service host-ports are much more useful than pod host-ports.  But happy to go either way on this one.  @thockin ?

**Comment by thockin** [comment]
Awesome find!

> Mini break-through (maybe?): It looks like we can create our ELBs in a different VPC subnet. And thus, we will know the IP range of the ELBs. So, for example, I created a second subnet of 172.20.1.0/24, and the source IP of traffic through ELB was 172.20.1.44. We still don't know which ELB is which, and thus we will still need a cluster-wide port assigned per externally-load-balanced-service (because ELB cannot target a different port per service), but I think we can avoid port collisions with kube-proxy's dynamic ports.

Could we create a subnet for each external service?  That way we could
tell which ELB was the source...

> My personal preference would still be to reserve a port range for these cluster-wide ports, and block kube-proxy and/or pod HostPorts from using them. I think that opens up two use-cases: cluster federation via SRV records and cheaper load-balancing by marking a service as externally available (without provisioning a load balancer). The only loss of functionality is for pod host-ports (which now can't use the reserved range), but I think service host-ports are much more useful than pod host-ports. But happy to go either way on this one. @thockin ?

For "cheaper load-balancing by marking a service as externally
available", this would get a random-ish port that was not the service
port, right?  I guess that is better than what we currently do with
PublicIPs, but it doesn't eliminate PublicIPs entirely.

Can you talk a bit more about how this plays with cluster federation?

**Comment by justinsb** [comment]
> Could we create a subnet for each external service?  That way we could tell which ELB was the source...

The default limit for subnets is surprisingly high (200 per VPC http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_Appendix_Limits.html).  I don't know how many people would actually run even 10 ELBs.  _Maybe_ we divide that 200 by 3 or 4 if people run multiple AZs in the same VPC, but it does seem like there is enough breathing room here.  It does feel like a hack, but it is an option.  Maybe it is good enough for V1.

But we don't get any of the fringe benefits of port assignment...

> Cheaper load-balancing

Yes, they would get a random-ish port by default.  We could think about letting people request a specific port in the same way that they request a specific hostport for a Pod.  But if people are using a CDN (CloudFlare?) maybe this is acceptable.  Or if they're using a client that understands SRV records, this may not be an issue.

> Cluster federation

I don't know whether this would actually work in practice, but what I was thinking is that if you had two clusters C1 and C2, and you wanted to consume a service from C2 in C1, the first thing you would have to do would be to expose that service from C1.  I was imagining that C1 would additionally expose a DNS server, publishing an SRV record for the service.  (I guess technically this wouldn't require a cluster-wide port, but it becomes really complicated without cluster-wide ports)

Then services in C2 could consume that SRV record:
* Directly (if it was a "real" domain name)
* or C2 could be told about C1's DNS server and bridge to it from C2's DNS (like a CNAME?)
* or C2's kube-proxy could know that some services were external, and instead of making a back-end connection to a pod, it could make a back-end connection to the results of an SRV resolve.

I don't know if any of these fit in to the current thinking on cluster federation.  I'm thinking about how I would run an HA cluster of mongodb servers, and this would be helpful for that scenario, probably consuming SRV records directly, maybe with the aid of a sidecar-process.

**Comment by thockin** [comment]
This is obsoleted by #7679 and friends.

**Pull request closed without merging**